medcat/utils/relation_extraction/rel_dataset.py

```diff
@@ -0,0 +1,503 @@
+
+                rows_to_remove = []
+                for row_idx in range(len(out_rels)):
+                    if len(out_rels[row_idx]["output_relations"]) < 1:
+                        rows_to_remove.append(row_idx)
+
+                relation_token_span_ids = []
+                out_ent1_ent2_starts = []
+
+                for rel in out_rels:
+                    if len(rel["output_relations"]) > 0:
+                        relation_token_span_ids.append(
+                            rel["output_relations"][0][0])
+                        out_ent1_ent2_starts.append(
+                            rel["output_relations"][0][1])
+                    else:
+                        relation_token_span_ids.append([])
+                        out_ent1_ent2_starts.append([])
+
+                df["relation_token_span_ids"] = relation_token_span_ids
+                df["ent1_ent2_start"] = out_ent1_ent2_starts
+
+                df = df.drop(index=rows_to_remove)
+                df = df.drop(columns=col)
+                break
+
+        nclasses, labels2idx, idx2label = RelData.get_labels(
+            df["label"], self.config)
+
+        output_relations = df.values.tolist()
+
+        print("No. of relations detected:", len(output_relations),
+              " from : ", csv_path, "classes: ", str(idx2label))
+
+        # replace/update label_id with actual detected label number
+        for idx in range(len(output_relations)):
+            output_relations[idx][5] = labels2idx[output_relations[idx][4]]
+
+        return {"output_relations": output_relations, "nclasses": nclasses, "labels2idx": labels2idx, "idx2label": idx2label}
+
+    def create_base_relations_from_doc(self, doc: Union[Doc, str], doc_id: str, ent1_ent2_tokens_start_pos: Union[List, Tuple] = (-1, -1)) -> Dict:
+        """  
+            doc : SpacyDoc
+            window_size : int, Character distance between any two entities start positions.
+            Creates a list of tuples based on pairs of entities detected (relation, ent1, ent2) for one spacy document.
+
+            Returns:
+                relations (List[list]) :
+                   data columns: [
+                                  "relation_token_span_ids", "ent1_ent2_start", "ent1", "ent2", "label",
+                                  "label_id", "ent1_type", "ent2_type", "ent1_id", "ent2_id",
+                                  "ent1_cui", "ent2_cui", "doc_id"
+                                 ]
+        """
+        relation_instances = []
+
+        chars_to_exclude = ":!@#$%^&*()-+?_=.,;<>/[]{}"
+        tokenizer_data = None
+
+        if isinstance(doc, str):
+            tokenizer_data = self.tokenizer(doc, truncation=False)
+            doc_text = doc
+        elif isinstance(doc, Doc):
+            tokenizer_data = self.tokenizer(doc.text, truncation=False)
+            doc_text = doc.text
+
+        doc_length = len(tokenizer_data["tokens"])
+
+        if ent1_ent2_tokens_start_pos != (-1, -1):
```

Comment:
Perhaps separate these if-else logics to separate methods?

Reply:
yeah agree - can these two live in different methods?

The logic is fundamentally the same and returned relation_id, if its directly a string or `spacy.Doc`. Can you transform the `spacy.Doc` into a string then re-use the same function logic to generate the tuple of entity / relations.

I'd struggle picking apart this logic spareated across these two large if / else branches